src/backup.py

```diff
@@ -0,0 +1,133 @@
+
+    @validator("region", always=True)
+    @classmethod
+    def check_region_or_endpoint_set(cls, region: str, values: dict[str, Any]) -> str:
+        """Validate that eather that region or endpoint are set.
+
+        Args:
+            region: region attribute
+            values: all attributes in S3 configuration
+
+        Returns:
+            value of the region attribute
+
+        Raises:
+            ValueError: if the configuration is invalid.
+        """
+        endpoint = values.get("endpoint")
+        if not region and not endpoint:
+            raise ValueError('one of "region" or "endpoint" needs to be set')
+        return region
+
+
+class SynapseBackup(Object):
+    """Class to manage Synapse backups over S3."""
+
+    _S3_RELATION_NAME = "s3-backup-parameters"
+
+    def __init__(self, charm: ops.CharmBase):
+        """Initialize the backup object.
+
+        Args:
+            charm: The parent charm the backups are made for.
+        """
+        super().__init__(charm, "backup")
+
+        self._charm = charm
+        self._s3_client = S3Requirer(self._charm, self._S3_RELATION_NAME)
+        self.framework.observe(
+            self._s3_client.on.credentials_changed, self._on_s3_credential_changed
+        )
+        self.framework.observe(self._s3_client.on.credentials_gone, self._on_s3_credential_gone)
+
+    def _on_s3_credential_changed(self, _: CredentialsChangedEvent) -> None:
+        """Check new S3 credentials set the unit to blocked if they are wrong."""
+        try:
+            s3_parameters = S3Parameters(**self._s3_client.get_s3_connection_info())
+        except ValueError:
+            self._charm.unit.status = ops.BlockedStatus(S3_INVALID_CONFIGURATION)
+            return
+
+        if not can_use_bucket(s3_parameters):
+            self._charm.unit.status = ops.BlockedStatus(S3_CANNOT_ACCESS_BUCKET)
+        logger.info("S3 backup correctly configured")
+
+    def _on_s3_credential_gone(self, _: CredentialsChangedEvent) -> None:
+        """Handle s3 credentials gone to reset unit status if it is now correct."""
+        if (
+            isinstance(self._charm.unit.status, ops.BlockedStatus)
+            and self._charm.unit.status.message in BACK_UP_STATUS_MESSAGES
+        ):
+            self._charm.unit.status = ops.ActiveStatus()
+
+
+def can_use_bucket(s3_parameters: S3Parameters) -> bool:
+    """Check if a bucket exists and is accessible in an S3 compatible object store.
+
+    Args:
+        s3_parameters: S3 connection parameters
+
+    Returns:
+       True if the bucket exists and is accessible
+    """
+    session = boto3.session.Session(
+        aws_access_key_id=s3_parameters.access_key,
+        aws_secret_access_key=s3_parameters.secret_key,
+        region_name=s3_parameters.region,
+    )
+    try:
+        s3 = session.resource("s3", endpoint_url=s3_parameters.endpoint)
+    except BotoCoreError:
+        logger.exception("Failed to create S3 session")
```

Comment:
It might be helpful to print the exception trace too

Reply:
`logger.exception` already logs the exception trace ([example execution](https://github.com/canonical/synapse-operator/actions/runs/7568727201/job/20612993165#step:13:518)). Not sure if that is what you are asking.